<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Homework #1 &mdash; cscb2022  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Homework #2" href="HW_02.html" />
    <link rel="prev" title="How to do RNA velcoity" href="../day/13_scvelo.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> cscb2022
            <img src="../_static/CSCB_logo2.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Schedule</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../schedule.html">Schedule overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../schedule.html#homework-and-final-project-due-dates">Homework and final project due dates</a></li>
<li class="toctree-l1"><a class="reference internal" href="../schedule.html#lecture-schedule">Lecture schedule</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Lectures:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../day/01_history_of_stem_cells.html">History of Stem Cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/02_introduction_to_python_part_1.html">Comp Lesson 1: Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/03_introduction_to_python_part_2.html">Comp Lesson 2: Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/04_Till_McCulloch.html">Till and McCulloch</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/05_scRNAseq_data_generation.html">scRNAseq data generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/06_scRNAseq_analysis.html">scRNAseq analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/07_scRNAseq_analysis_part_2.html">scRNAseq analysis part 2</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/08_cell_identity.html">Cell identity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/09_trajectory_inference.html">Trajectory inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/10_TI_part_2.html">How to do TI analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/11_stemness.html">How to do ‘Stemness’ analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/12_RNA_velocity.html">RNA Velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../day/13_scvelo.html">How to do RNA velcoity</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Homework</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Homework #1</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Goal:-Ensure-that-you-…">Goal: Ensure that you …</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Preamble:-functions">Preamble: functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Part-1:-Write-a-function-prose_generator.">Part 1: Write a function <code class="docutils literal notranslate"><span class="pre">prose_generator</span></code>.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Part-2:-Improve-your-function">Part 2: Improve your function</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Part-3:-Use-Pandas-(+-Numpy),-and-Seaborn-(+-Matplotlib)">Part 3: Use Pandas (+ Numpy), and Seaborn (+ Matplotlib)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#3A:-Load-the-data">3A: Load the data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3B:-Filter-out-genes-whose-maximum-expression-&lt;-2">3B: Filter out genes whose maximum expression &lt; 2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3C:-Make-a-box-and-whiskers-plot-of-the-gene-Pou5f1,-grouped-by-cell/tissue-type.">3C: Make a box and whiskers plot of the gene Pou5f1, grouped by cell/tissue type.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3D:-Determine-the-top-5-genes-expressed-per-cell/tissue-type">3D: Determine the top 5 genes expressed per cell/tissue type</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3E:-Compute-the-similarity-between-samples.">3E: Compute the similarity between samples.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Part-4:-Modeling-Birth-and-death-of-CFUs">Part 4: Modeling Birth and death of CFUs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Part-4A:-Write-a-function-to-simulate-the-number-of-colony-forming-units.">Part 4A: Write a function to simulate the number of colony forming units.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Part-4B:-Compare-in-silico-experiments-to-empirical-data">Part 4B: Compare <em>in silico</em> experiments to empirical data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="HW_02.html">Homework #2</a></li>
<li class="toctree-l1"><a class="reference internal" href="HW_03.html">Homework #3</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cscb2022</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Homework #1</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/homework/HW_01.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Homework-#1">
<h1>Homework #1<a class="headerlink" href="#Homework-#1" title="Permalink to this headline"></a></h1>
<p>Due 11:59 pm EST, <strong>Sunday Feb 13th, 2022</strong> Email your solutions (both .ipnb and .html files) to: <a class="reference external" href="mailto:compscbio&#37;&#52;&#48;gmail&#46;com">compscbio<span>&#64;</span>gmail<span>&#46;</span>com</a>.</p>
<section id="Goal:-Ensure-that-you-…">
<h2>Goal: Ensure that you …<a class="headerlink" href="#Goal:-Ensure-that-you-…" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>have basic, practical knowledge of Python</p></li>
<li><p>are able to install Python packages</p></li>
<li><p>are able to use and create Jupyter notebooks</p></li>
<li><p>can apply this knowledge to explore one aspect of self-renewal</p></li>
</ul>
<hr></section>
<section id="Preamble:-functions">
<h2>Preamble: functions<a class="headerlink" href="#Preamble:-functions" title="Permalink to this headline"></a></h2>
<p>In class, we did not cover the syntax for defining your own functions. Let’s cover this now. Note that we use the <code class="docutils literal notranslate"><span class="pre">randint</span></code> function from the random module, and so need to import this module. Typically, you should import all of the modules that you will need at the start of your session. So, I am going to put some other import statements here that we won’t need until later in the homework. Please install modules, as needed, using either <code class="docutils literal notranslate"><span class="pre">pip</span></code> or anaconda.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">nltk</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
<p>Here is a funky little function. It just prints a string passed to the function a random number of times between 1 and 10.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">myFirstFunction</span><span class="p">(</span><span class="n">arg1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;This is just an example; it doesn&#39;t do anything useful&quot;&quot;&quot;</span>
    <span class="n">times_to_repeat</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">times_to_repeat</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">arg1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">times_to_repeat</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">myFirstFunction</span><span class="p">(</span><span class="s2">&quot;howdy&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
howdy
howdy
howdy
howdy
howdy
howdy
howdy
howdy
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
8
</pre></div></div>
</div>
<p>When you create a function, you need to use the <code class="docutils literal notranslate"><span class="pre">def</span></code> keyword, then supply your function name, followed by the names of arguments that need to be passed to the function enclosed in parenthesis, followed by <code class="docutils literal notranslate"><span class="pre">:</span></code>. Then the following indented suite of code is where you define what the function will do. The first line should be text enclosed in three pairs of double quotes. This is what will be returned when <code class="docutils literal notranslate"><span class="pre">help(yourFunction)</span></code> is invoked. Therefore, give some thought to this text and make it
a pithy description of your function. If you do not end the function suite with an explicit call to <code class="docutils literal notranslate"><span class="pre">return</span></code>, the the function will by default return <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<hr></section>
<section id="Part-1:-Write-a-function-prose_generator.">
<h2>Part 1: Write a function <code class="docutils literal notranslate"><span class="pre">prose_generator</span></code>.<a class="headerlink" href="#Part-1:-Write-a-function-prose_generator." title="Permalink to this headline"></a></h2>
<p>The first thing that you must do for this homeowrk is to define a function that writes a number of English sentences. The user will be asked to specify the number of sentences. The function will also need to take in as input lists of nouns and predicates. The function will construct and print out sentences of the pattern: (noun, predicate), by randomly selecting words from the corresponding lists. Below, we get you started with the skeleton of the function.</p>
<p>To get full credit for Part 1, your answer must include a function that takes as input the number of sentences to create, and lists of user-supplied nouns and predicates. The lists can be entered either by prompting the user, or by specifying a file(s) that contains the words. If the former, define your examples in your notebook. If the latter, inlcude them as supplemental files with your submission. Your function must randomly select from these lists to make sentences that start with the noun
and end with the predicate. Finally, your function must output the specified number of sentences.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Part 1: modify the function skeleton below to meet the conditions stated above</span>

<span class="kn">import</span> <span class="nn">random</span>

<span class="k">def</span> <span class="nf">prose_generator</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot; a function that writes [beautiful] prose &quot;&quot;&quot;</span>
    <span class="c1"># put your code here</span>

    <span class="k">return</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Part 1: Show us an example of running your function here</span>
</pre></div>
</div>
</div>
<HR></section>
<section id="Part-2:-Improve-your-function">
<h2>Part 2: Improve your function<a class="headerlink" href="#Part-2:-Improve-your-function" title="Permalink to this headline"></a></h2>
<p>Your function probably does not have much error checking so far. And, it is possible that a random sentence will be generated more than once. I don’t know about you, but I find it mildly annoying when people say the same thing repeatedly. So, your second task is to modify your function (or create a new one) so that the following conditions are met:</p>
<ol class="upperalpha simple">
<li><p>That the nouns are actually nouns.</p></li>
<li><p>That at least two valid nouns and two predicates are entered</p></li>
<li><p>That a sentence is printed at most once. In other words, prevent the function from outputing duplicate sentences.</p></li>
</ol>
<p>To achieve (A), you can compare the user-supplied nouns against the long list of nouns that we show you how to extract from the NLTK resource below. If the noun supplied does not occur in this list, then your function must at the very least, notify the user. To get full credit for Part 2, your new or modified sentence-generator function must do the same things as in Part 1, plus it must meet the conditions (A-C).</p>
<p>You will need to install the following resources from nltk (see next code cell). For more information about the NLTK resource: <a class="reference external" href="https://www.nltk.org/data.html">https://www.nltk.org/data.html</a></p>
<p>The post that we found useful to get a list of lots of Enlgish nouns: <a class="reference external" href="https://stackoverflow.com/questions/17753182/getting-a-large-list-of-nouns-or-adjectives-in-python-with-nltk-or-python-mad">https://stackoverflow.com/questions/17753182/getting-a-large-list-of-nouns-or-adjectives-in-python-with-nltk-or-python-mad</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nltk</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s1">&#39;omw-1.4&#39;</span><span class="p">)</span>
<span class="n">nltk</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s2">&quot;wordnet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     /Volumes/forest/moved_users/pcahan1/nltk_data...
[nltk_data]   Unzipping corpora/omw-1.4.zip.
[nltk_data] Downloading package wordnet to
[nltk_data]     /Volumes/forest/moved_users/pcahan1/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">nltk.corpus</span> <span class="kn">import</span> <span class="n">wordnet</span> <span class="k">as</span> <span class="n">wn</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_nouns</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">synset</span> <span class="ow">in</span> <span class="n">wn</span><span class="o">.</span><span class="n">all_synsets</span><span class="p">(</span><span class="s1">&#39;n&#39;</span><span class="p">):</span>
    <span class="n">all_nouns</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">synset</span><span class="o">.</span><span class="n">lemma_names</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">all_nouns</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
146347
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="s2">&quot;carrot&quot;</span> <span class="ow">in</span> <span class="n">all_nouns</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="s2">&quot;faba-laba&quot;</span> <span class="ow">in</span> <span class="n">all_nouns</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
False
</pre></div></div>
</div>
<HR></section>
<section id="Part-3:-Use-Pandas-(+-Numpy),-and-Seaborn-(+-Matplotlib)">
<h2>Part 3: Use Pandas (+ Numpy), and Seaborn (+ Matplotlib)<a class="headerlink" href="#Part-3:-Use-Pandas-(+-Numpy),-and-Seaborn-(+-Matplotlib)" title="Permalink to this headline"></a></h2>
<p>Next week (actually this week), we are going to start using Scanpy to analyze scRNAseq data. The Scanpy package is excellent at making it easy to analyze this type of data. But when you are just getting started, this can almost be <em>too</em> easy because it hides a lot of important details. The goal of this section is to expose you to some aspects of cleaning and selecting data, and to some basic visualization functions. This will give you a better idea of what is happening under-the-hood with
packages such as Scanpy (or Seurat in R).</p>
<p>In this section, you need to:</p>
<ul class="simple">
<li><ol class="upperalpha simple">
<li><p>Load a gene expression data set (<a class="reference external" href="https://compstemcellbio.s3.amazonaws.com/2022/homework/01/hw1_gex_df.csv">expression table</a>; <a class="reference external" href="https://compstemcellbio.s3.amazonaws.com/2022/homework/01/hw1_gex_sampTab.csv">meta-data</a>)</p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="2">
<li><p>Remove genes that are not expressed above a selected threshold</p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="3">
<li><p>Create boxplot of the expression of one of my favorite genes, Pou5f1, across cell/tissue types</p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="4">
<li><p>Identify the top 5 genes in terms of mean expression per cell/tissue type &amp; plot a heatmap of these genes (ordered)</p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="5">
<li><p>Compute the similarity of samples &amp; plot a heatmap of them (ordered)</p></li>
</ol>
</li>
</ul>
<p>To do this somewhat efficiently, you might want to look at the following documentation:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html">https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.max.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.max.html</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html">https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html</a></p></li>
<li><p><a class="reference external" href="https://seaborn.pydata.org/generated/seaborn.boxplot.html">https://seaborn.pydata.org/generated/seaborn.boxplot.html</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rank.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rank.html</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html</a></p></li>
<li><p><a class="reference external" href="https://seaborn.pydata.org/introduction.html">https://seaborn.pydata.org/introduction.html</a></p></li>
<li><p><a class="reference external" href="https://seaborn.pydata.org/generated/seaborn.heatmap.html">https://seaborn.pydata.org/generated/seaborn.heatmap.html</a></p></li>
</ul>
<p>If it sounds like a lot, don’t worry it will be fun!</p>
<section id="3A:-Load-the-data">
<h3>3A: Load the data<a class="headerlink" href="#3A:-Load-the-data" title="Permalink to this headline"></a></h3>
<p>Download the expression matrix and meta-data from the links above. Use the Pandas <code class="docutils literal notranslate"><span class="pre">read_csv()</span></code> to load these. This will create a Pandas DataFrame.</p>
<p>This mouse data comes from one of our papers, where we compiled previously published RNA-Seq data for classification purposes. There are 16 cell/tissue types, each represented by approximately 60 or more individual samples. Later, in Step 3C, you should also load the metadata, which indicates the cell/tissue type for each sample.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 3A: Enter your code for loading the file hw1_df.csv as a Data Frame and displaying part of it.</span>
</pre></div>
</div>
</div>
</section>
<section id="3B:-Filter-out-genes-whose-maximum-expression-&lt;-2">
<h3>3B: Filter out genes whose maximum expression &lt; 2<a class="headerlink" href="#3B:-Filter-out-genes-whose-maximum-expression-<-2" title="Permalink to this headline"></a></h3>
<p>You might consider using the <code class="docutils literal notranslate"><span class="pre">pd.loc</span></code> and <code class="docutils literal notranslate"><span class="pre">pd.max()</span></code> functions</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[138]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 3B: Enter your code for removing genes and print out the size of the resulting data frame</span>
</pre></div>
</div>
</div>
<p>Limit genes to those that have maximum expression &gt; 2</p>
</section>
<section id="3C:-Make-a-box-and-whiskers-plot-of-the-gene-Pou5f1,-grouped-by-cell/tissue-type.">
<h3>3C: Make a box and whiskers plot of the gene Pou5f1, grouped by cell/tissue type.<a class="headerlink" href="#3C:-Make-a-box-and-whiskers-plot-of-the-gene-Pou5f1,-grouped-by-cell/tissue-type." title="Permalink to this headline"></a></h3>
<p>To do this you will need to load the meta-data, linked to above. You might want to use the <code class="docutils literal notranslate"><span class="pre">pd.concat()</span></code> and <code class="docutils literal notranslate"><span class="pre">sns.boxplot()</span></code> functions.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 3C: Show you code for doing what you need to do to ultimatley make the boxplot.</span>
<span class="c1"># And, of course, display the boxplot below, too</span>
</pre></div>
</div>
</div>
</section>
<section id="3D:-Determine-the-top-5-genes-expressed-per-cell/tissue-type">
<h3>3D: Determine the top 5 genes expressed per cell/tissue type<a class="headerlink" href="#3D:-Determine-the-top-5-genes-expressed-per-cell/tissue-type" title="Permalink to this headline"></a></h3>
<p>Once you have done so, make a heatmap showing expression of these genes across all samples. To get full credit for this, you need to</p>
<ul class="simple">
<li><p>exclude duplicate genes (i.e. genes that are in the top 5 in more than one cell/tissue type)</p></li>
<li><p>order the samples by cell/tissue type</p></li>
</ul>
<p>You might want to use the Pandas functions <code class="docutils literal notranslate"><span class="pre">groupby</span></code> and <code class="docutils literal notranslate"><span class="pre">mean</span></code> and <code class="docutils literal notranslate"><span class="pre">sort_values</span></code> here, and it may be useful to do some type conversions. Do use <code class="docutils literal notranslate"><span class="pre">sns.heatmap()</span></code> for plotting.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[140]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 3D: Put your code for this part here.</span>
</pre></div>
</div>
</div>
</section>
<section id="3E:-Compute-the-similarity-between-samples.">
<h3>3E: Compute the similarity between samples.<a class="headerlink" href="#3E:-Compute-the-similarity-between-samples." title="Permalink to this headline"></a></h3>
<p>Compute the Pearson correaltion coefficient between all pairs of samples. Plot this as a heatmap. Make sure the samples are ordered by cell/tissue type. You might want to use the Pandas <code class="docutils literal notranslate"><span class="pre">corr()</span></code> function. Be careful of what your are computing correlation of.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 3E: Put you code for 3E here</span>
</pre></div>
</div>
</div>
<hr></section>
</section>
<section id="Part-4:-Modeling-Birth-and-death-of-CFUs">
<h2>Part 4: Modeling Birth and death of CFUs<a class="headerlink" href="#Part-4:-Modeling-Birth-and-death-of-CFUs" title="Permalink to this headline"></a></h2>
<p>Recall the self-renewal assay from Siminovitch, Till and McCulloch 1963 (right) <img alt="e7c501514ef34986b8c6725feee71c4d" class="no-scaled-link" src="../_images/CFUs_selfrenewal.png" style="width: 200px;" /> In this assay, nodules were dissected, cells were dispersed, resuspended, and tranplanted into secondary recipients. Then after 10 days (or later depending on the experiment), the number of spleen nodules were counted per recipient. Recall that the number of nodules (or colonies) per spleen was highly heterogeneous. To the authors, this indicated that heterogeneity was “… a feature
either of colony-forming cells or of the process of colony-formation.”</p>
<p><img alt="1b92c5f720fe4779a6cc00da0525c723" class="no-scaled-link" src="../_images/birth_death.png" style="width: 200px;" /> In a follow-up <a class="reference external" href="https://blackboard.jhu.edu/bbcswebdav/pid-11413450-dt-content-rid-113157130_2/xid-113157130_2">paper: Till, McCulloch, and Siminovitch, 1963</a>, this trio proposed that the heterotengeity could be due to inherent stochasticity in each colony-forming cell. In other words, a colony forming cell “… either give[s] rise to progeny like itself (“birth”), or [is] removed in some way (“death”), and these two events occur in a random fashion.” This is
often referred to as a “Birth-and-death process”. They reasoned that a cell will give birth to another colony forming cell with Probability = p2 or to a differentiated cells with Probability = p0 (see panel at left). They further reasoned that if this model is correct, then simulations using the model must be aligned with their empircal observations.</p>
<p>The goal of Part 4 is for you to replicate some of what the authors did to test this reasoning:</p>
<blockquote>
<div><p>A test of this assumption was made by the use of the Monte Carlo method. In this method, the choice between occurrence or nonoccurrence of a random event is made by drawing a number from a table of random digits. Consider as an example the case in which the birth and death probabilities, p2 and p0, are arbitrarily set equal to 0.6 and 0.4, respectively. Let the six digits, 0, 1, 2, 3, 4, and 5 signify a “birth” and the four digits 6, 7, 8, and 9 signify a “death.” If the first random number
to be drawn was a 5, then the model would show a birth, or an increase from one cell to two cells. Each subsequent choice of a number would determine the fate of a cell in a similar manner.</p>
</div></blockquote>
<section id="Part-4A:-Write-a-function-to-simulate-the-number-of-colony-forming-units.">
<h3>Part 4A: Write a function to simulate the number of colony forming units.<a class="headerlink" href="#Part-4A:-Write-a-function-to-simulate-the-number-of-colony-forming-units." title="Permalink to this headline"></a></h3>
<p>Your task is to precisley mimic what the authors state above: write a function that takes as input:</p>
<ul class="simple">
<li><p>the number of time steps, or generations</p></li>
<li><p>P2 = probability of birth, resulting in two colony forming units</p></li>
</ul>
<p>and that outputs the number of CFUs remaining. Note that a death event results in a loss of one CFU, whereas a birth yields two CFUs. You may consider using the Numpy function <code class="docutils literal notranslate"><span class="pre">random.uniform()</span></code> here.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 4A: Write your function definition here</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 4A: Show us an example of how it works here</span>
</pre></div>
</div>
</div>
</section>
<section id="Part-4B:-Compare-in-silico-experiments-to-empirical-data">
<h3>Part 4B: Compare <em>in silico</em> experiments to empirical data<a class="headerlink" href="#Part-4B:-Compare-in-silico-experiments-to-empirical-data" title="Permalink to this headline"></a></h3>
<p>Use your function to determine how values of P2 influence the distribution of CFUs, and how this compares to real data. To do this, you will need to call your function many times. For this homework, let’s stay close to the original paper (n=52 recipient spleens). You will also need the original <a class="reference external" href="https://compstemcellbio.s3.amazonaws.com/2022/homework/01/hw1_cfu_data.csv">data</a>, which you can load with the usual <code class="docutils literal notranslate"><span class="pre">pd.read_csv()</span></code> function. We recommend testing 3 - 5 values for P2. As you will
see from the original paper, the simulated and experimentally derived data were compared by examining cumulative distribution functions. You can readily plot these using the Seaborn <code class="docutils literal notranslate"><span class="pre">ecdfplot()</span></code> function.</p>
<p>Bonus points: Compare these distrubitions to the Gamma distribution mentioned in the paper.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 4B: Show your code and figure(s) here.</span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../day/13_scvelo.html" class="btn btn-neutral float-left" title="How to do RNA velcoity" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="HW_02.html" class="btn btn-neutral float-right" title="Homework #2" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Patrick Cahan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-LFQLTP4DZG"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-LFQLTP4DZG', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>