<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>&lt;no title&gt; &mdash; cscb2022  documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/underscore.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../../../index.html" class="icon icon-home"> cscb2022
            <img src="../../../../_static/CSCB_logo2.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Schedule</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../schedule.html">Schedule overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../schedule.html#homework-and-final-project-due-dates">Homework and final project due dates</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../schedule.html#lecture-schedule">Lecture schedule</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Lectures:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/01_history_of_stem_cells.html">History of Stem Cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/02_introduction_to_python_part_1.html">Comp Lesson 1: Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/03_introduction_to_python_part_2.html">Comp Lesson 2: Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/04_Till_McCulloch.html">Till and McCulloch</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/05_scRNAseq_data_generation.html">scRNAseq data generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/06_scRNAseq_analysis.html">scRNAseq analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/07_scRNAseq_analysis_part_2.html">scRNAseq analysis part 2</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/08_cell_identity.html">Cell identity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/09_trajectory_inference.html">Trajectory inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/10_TI_part_2.html">How to do TI analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/11_stemness.html">How to do ‘Stemness’ analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/12_RNA_velocity.html">RNA Velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/13_scvelo.html">RNA velocity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/14_Lineage_tracing.html">Lineage tracing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/15_Pluripotency.html">Pluripotency</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../day/16_Spatial_Cell-to-Cell.html">Spatial transcriptomics &amp; Cell-to-cell communication</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Homework</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../homework/HW_01.html">Homework #1</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../homework/HW_02.html">Homework #2</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../homework/HW_03.html">Homework #3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../homework/HW_04.html">Homework #4</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../homework/HW_05.html">Homework #5</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../homework/HW_06.html">Homework #6</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">cscb2022</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>&lt;no title&gt;</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../../_sources/docs/doctrees/nbsphinx/day/17_Lineage_Tracing.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<p>Today, we will analyze sc lineage tracing data with <a class="reference external" href="https://www.nature.com/articles/s41587-022-01209-1">CoSpar Wang et al 2022</a></p>
<p>Much of the tutorial today comes straight from <a class="reference external" href="https://cospar.readthedocs.io/en/latest/index.html">CoSpar’s documentation</a></p>
<figure><p><img alt="afbffea208524cd9b2350b7ecb2a1a72" src="docs/doctrees/nbsphinx/day/img/cospar_schematic.png" /></p>
<figcaption><p>CoSpar Overview. The transition matrix T computed by maximum likelihood estimation.</p>
</figcaption><div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;/figure&gt;
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!pip3 install cospar</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cospar</span> <span class="k">as</span> <span class="nn">cs</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span><span class="o">=</span><span class="mi">2</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Data from <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/31974159/">Weinreb et al Science 2020</a>.</p>
<figure><p><img alt="07009ecc73b64a6c9f0aec85c1453145" src="docs/doctrees/nbsphinx/day/img/hspc_weinreb_schematic.png" /></p>
<figcaption><p>Data collection strategy.</p>
</figcaption><div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;/figure&gt;
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_orig</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;adata_cospar_hspc.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Required slots in annData for CoSpar:</p>
<ul class="simple">
<li><p>adata.X: state count matrix, shape (n_cell, n_gene). This should not be log-normalized.</p></li>
<li><p>adata.var_names: list of gene names, shape (n_genes,).</p></li>
<li><p>adata.obs[‘time_info’]: time annotation (type: str) for each cell, shape (n_cell,).</p></li>
<li><p>adata.obs[‘state_info’]: state annotation for each cell, shape (n_cell, 1). [Optional. Can be generated in preprocessing.</p></li>
<li><p>adata.obsm[‘X_clone’]: clonal labels for each cell in the form of np.array or sparse matrix, shape (n_cell, n_clone).</p></li>
<li><p>adata.obsm[‘X_pca’]: PCA matrix, shape (n_cell, n_pcs). [Optional. Can be generated in preprocessing.</p></li>
<li><p>adata.obsm[‘X_emb’]: two-dimensional embedding, shape (n_cell, 2). [Optional. Can be generated in preprocessing.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">set_up_folders</span><span class="p">()</span> <span class="c1"># NB we did not run this in lecture but will be necessary</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">NameError</span>                                 Traceback (most recent call last)
<span class="ansi-green-fg">/var/folders/24/v4g076gx0mz2mwtd6ktzw8_h0000gn/T/ipykernel_45011/1119742059.py</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-fg">----&gt; 1</span><span class="ansi-red-fg"> </span>cs<span class="ansi-blue-fg">.</span>hf<span class="ansi-blue-fg">.</span>set_up_folders<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span> <span class="ansi-red-fg"># NB we did not run this in lecture but will be necessary</span>

<span class="ansi-red-fg">NameError</span>: name &#39;cs&#39; is not defined
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_orig</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 7438 × 25289
    obs: &#39;time_info&#39;, &#39;state_info&#39;
    uns: &#39;clonal_time_points&#39;, &#39;data_des&#39;, &#39;state_info_colors&#39;, &#39;time_info_colors&#39;, &#39;time_ordering&#39;
    obsm: &#39;X_clone&#39;, &#39;X_emb&#39;, &#39;X_pca&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;state_info&#39;</span><span class="p">,</span> <span class="n">projection</span><span class="o">=</span><span class="s1">&#39;3d&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_9_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_9_0.png" style="width: 442px; height: 354px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;emb&quot;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;state_info&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_10_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_10_0.png" style="width: 627px; height: 366px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;emb&quot;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_11_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_11_0.png" style="width: 509px; height: 366px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;emb&quot;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;time_info&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_12_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_12_0.png" style="width: 556px; height: 366px;" />
</div>
</div>
<p>We can examine the location of clones on embeddings</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clones_on_manifold</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clone_list</span><span class="o">=</span><span class="p">[</span><span class="mi">151</span><span class="p">],</span><span class="n">color_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;blue&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_14_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_14_0.png" style="width: 330px; height: 306px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clones_on_manifold</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clone_list</span><span class="o">=</span><span class="p">[</span><span class="mi">200</span><span class="p">],</span><span class="n">color_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;blue&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_15_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_15_0.png" style="width: 330px; height: 306px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clones_on_manifold</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clone_list</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">color_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;blue&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_16_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_16_0.png" style="width: 330px; height: 306px;" />
</div>
</div>
<p>How often are clones shared across cell types? This analysis done for given time point</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_times</span><span class="o">=</span><span class="s1">&#39;4&#39;</span>
<span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;X_clone&#39;</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span><span class="n">selected_times</span><span class="o">=</span><span class="n">selected_times</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;X_clone&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as dictionary at adata.uns[&#39;fate_coupling_X_clone&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:title={&#39;center&#39;:&#39;source: X_clone&#39;}&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_18_2.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_18_2.png" style="width: 435px; height: 350px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_times</span><span class="o">=</span><span class="s1">&#39;6&#39;</span>
<span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;X_clone&#39;</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span><span class="n">selected_times</span><span class="o">=</span><span class="n">selected_times</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;X_clone&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as dictionary at adata.uns[&#39;fate_coupling_X_clone&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:title={&#39;center&#39;:&#39;source: X_clone&#39;}&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_19_2.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_19_2.png" style="width: 425px; height: 350px;" />
</div>
</div>
<p>We can look at barcode distributions across cell types, too</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_times</span><span class="o">=</span><span class="s1">&#39;4&#39;</span>
<span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barcode_heatmap</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_times</span><span class="o">=</span><span class="n">selected_times</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data saved at adata.uns[&#39;barcode_heatmap&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_21_2.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_21_2.png" style="width: 377px; height: 577px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_multitime_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">clonal_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">],</span><span class="n">later_time_point</span><span class="o">=</span><span class="s1">&#39;6&#39;</span><span class="p">,</span>
        <span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span><span class="n">sparsity_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">max_iter_N</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Trying to set attribute `._uns` of view, copying.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
------Compute the full Similarity matrix if necessary------
------Infer transition map between initial time points and the later time one------
--------Current initial time point: 2--------
Step 1: Select time points
Number of multi-time clones post selection: 185
Step 2: Optimize the transition map recursively
Load pre-computed similarity matrix
Iteration 1, Use smooth_round=20
Iteration 2, Use smooth_round=15
Iteration 3, Use smooth_round=10
Convergence (CoSpar, iter_N=3): corr(previous_T, current_T)=0.953
--------Current initial time point: 4--------
Step 1: Select time points
Number of multi-time clones post selection: 500
Step 2: Optimize the transition map recursively
Load pre-computed similarity matrix
Iteration 1, Use smooth_round=20
Iteration 2, Use smooth_round=15
Iteration 3, Use smooth_round=10
Convergence (CoSpar, iter_N=3): corr(previous_T, current_T)=0.974
-----------Total used time: 19.995919227600098 s ------------
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 7438 × 25289
    obs: &#39;time_info&#39;, &#39;state_info&#39;
    uns: &#39;clonal_time_points&#39;, &#39;data_des&#39;, &#39;state_info_colors&#39;, &#39;time_info_colors&#39;, &#39;time_ordering&#39;, &#39;available_map&#39;, &#39;fate_coupling_X_clone&#39;, &#39;barcode_heatmap&#39;, &#39;Tmap_cell_id_t1&#39;, &#39;Tmap_cell_id_t2&#39;, &#39;clonal_cell_id_t1&#39;, &#39;clonal_cell_id_t2&#39;, &#39;sp_idx&#39;, &#39;transition_map&#39;, &#39;intraclone_transition_map&#39;
    obsm: &#39;X_clone&#39;, &#39;X_emb&#39;, &#39;X_pca&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;transition_map&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<p>Now, given an state, we can examine it’s likely trajectory using the transition matrix just computed</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                    <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backward</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_26_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_26_0.png" style="width: 378px; height: 328px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">100</span><span class="p">]</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                    <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backward</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_27_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_27_0.png" style="width: 378px; height: 328px;" />
</div>
</div>
<p>Similarly, we can compute transition profiles for end states (where did they originate)?</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                    <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backward</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_29_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_29_0.png" style="width: 378px; height: 328px;" />
</div>
</div>
<p>Now let’s examine a fate map. Given a cluster/state/end-point, where did it originate?</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
               <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backward</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span>
               <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">show_histogram</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved at adata.obs[&#39;fate_map_transition_map_Neutrophil&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Monocyte&#39;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_31_1.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_31_1.png" style="width: 468px; height: 327px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
               <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">show_histogram</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_32_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_32_0.png" style="width: 468px; height: 327px;" />
</div>
</div>
<p>Compute fate potency as the normalized sum of potential endpoints</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_potency</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">map_backward</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;norm-sum&#39;</span><span class="p">,</span><span class="n">fate_count</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_potency</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved at adata.obs[&#39;fate_map_transition_map_Erythroid&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Lymphoid&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Eos&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_pDC&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_undiff&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Neutrophil&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Monocyte&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Mast&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Baso&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Ccr7_DC&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Neu_Mon&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Meg&#39;]
Results saved at adata.obs[&#39;fate_potency_transition_map&#39;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_34_1.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_34_1.png" style="width: 469px; height: 328px;" />
</div>
</div>
<p>Now let’s examine fate bias between two cell types/states/endpoints a and b. Computed as Pr(endstate = a) / Pr(endstate=a) + Pr(endstate=b)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">pseudo_count</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">selected_times</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved at adata.obs[&#39;fate_map_transition_map_Neutrophil&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Monocyte&#39;]
Results saved at adata.obs[&#39;fate_bias_transition_map_Neutrophil*Monocyte&#39;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_36_1.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_36_1.png" style="width: 468px; height: 327px;" />
</div>
</div>
<p>Fate coupling from transition matrix</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as dictionary at adata.uns[&#39;fate_coupling_transition_map&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:title={&#39;center&#39;:&#39;source: transition_map&#39;}&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_38_2.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_38_2.png" style="width: 433px; height: 350px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># compared to from clonal data alone</span>
<span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;X_clone&#39;</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;X_clone&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as dictionary at adata.uns[&#39;fate_coupling_X_clone&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:title={&#39;center&#39;:&#39;source: X_clone&#39;}&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_39_2.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_39_2.png" style="width: 433px; height: 350px;" />
</div>
</div>
<p>Now reconstruct the hierarchy</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">fate_hierarchy</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_hierarchy</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as dictionary at adata.uns[&#39;fate_hierarchy_transition_map&#39;]

            /-Baso
         /-|
      /-|   \-Eos
     |  |
   /-|   \-Mast
  |  |
  |  |   /-Erythroid
  |   \-|
--|      \-Meg
  |
  |      /-Monocyte
  |   /-|
  |  |   \-Neutrophil
   \-|
     |      /-pDC
     |   /-|
      \-|   \-Ccr7_DC
        |
         \-Lymphoid
</pre></div></div>
</div>
<p>What distinguishes progenitors likely to become monocytes from those that are likely to become neutrophils?</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">progenitor</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
 <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span> <span class="n">map_backward</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">bias_threshold_A</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">bias_threshold_B</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">progenitor</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
 <span class="n">source</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved at adata.obs[&#39;fate_map_transition_map_Neutrophil&#39;]
Results saved at adata.obs[&#39;fate_map_transition_map_Monocyte&#39;]
Results saved at adata.obs[&#39;fate_bias_transition_map_Neutrophil*Monocyte&#39;]
Results saved at adata.obs[f&#39;progenitor_transition_map_Neutrophil&#39;] and adata.obs[f&#39;diff_trajectory_transition_map_Neutrophil&#39;]
Results saved at adata.obs[f&#39;progenitor_transition_map_Monocyte&#39;] and adata.obs[f&#39;diff_trajectory_transition_map_Monocyte&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_43_1.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_43_1.png" style="width: 378px; height: 328px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_43_2.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_43_2.png" style="width: 378px; height: 328px;" />
</div>
</div>
<p>note that resuls are stored as follows:</p>
<p><code class="docutils literal notranslate"><span class="pre">adata.obs[f'progenitor_{source}_{fate_name}']</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">adata.obs[f'diff_trajectory_{source}_{fate_name}']</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">cell_group_A</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;diff_trajectory_transition_map_Neutrophil&#39;</span><span class="p">])</span>
<span class="n">cell_group_B</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;diff_trajectory_transition_map_Monocyte&#39;</span><span class="p">])</span>
<span class="n">dge_gene_A</span><span class="p">,</span> <span class="n">dge_gene_B</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">differential_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">cell_group_A</span><span class="o">=</span><span class="n">cell_group_A</span><span class="p">,</span><span class="n">cell_group_B</span><span class="o">=</span><span class="n">cell_group_B</span><span class="p">,</span><span class="n">FDR_cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dge_gene_A</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index</th>
      <th>gene</th>
      <th>Qvalue</th>
      <th>mean_1</th>
      <th>mean_2</th>
      <th>ratio</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5</td>
      <td>Mmp8</td>
      <td>0.000000e+00</td>
      <td>0.563468</td>
      <td>11.657158</td>
      <td>-3.017132</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>Wfdc17</td>
      <td>0.000000e+00</td>
      <td>0.078693</td>
      <td>7.124163</td>
      <td>-2.912935</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9</td>
      <td>Lpl</td>
      <td>2.733934e-267</td>
      <td>0.240403</td>
      <td>8.100443</td>
      <td>-2.875127</td>
    </tr>
    <tr>
      <th>3</th>
      <td>10</td>
      <td>Fabp5</td>
      <td>1.209024e-246</td>
      <td>1.652049</td>
      <td>15.526230</td>
      <td>-2.639579</td>
    </tr>
    <tr>
      <th>4</th>
      <td>3</td>
      <td>Psap</td>
      <td>0.000000e+00</td>
      <td>1.660915</td>
      <td>14.049889</td>
      <td>-2.499758</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1067</th>
      <td>2060</td>
      <td>H3f3a</td>
      <td>7.928128e-04</td>
      <td>4.989668</td>
      <td>5.169685</td>
      <td>-0.042721</td>
    </tr>
    <tr>
      <th>1068</th>
      <td>3290</td>
      <td>AB124611</td>
      <td>4.108763e-02</td>
      <td>0.605892</td>
      <td>0.648909</td>
      <td>-0.038136</td>
    </tr>
    <tr>
      <th>1069</th>
      <td>3124</td>
      <td>Tmpo</td>
      <td>3.081903e-02</td>
      <td>0.922378</td>
      <td>0.929656</td>
      <td>-0.005451</td>
    </tr>
    <tr>
      <th>1070</th>
      <td>2697</td>
      <td>Ffar2</td>
      <td>1.080521e-02</td>
      <td>0.608486</td>
      <td>0.609789</td>
      <td>-0.001169</td>
    </tr>
    <tr>
      <th>1071</th>
      <td>2857</td>
      <td>Nom1</td>
      <td>1.716971e-02</td>
      <td>0.235725</td>
      <td>0.236709</td>
      <td>-0.001149</td>
    </tr>
  </tbody>
</table>
<p>1072 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dge_gene_B</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index</th>
      <th>gene</th>
      <th>Qvalue</th>
      <th>mean_1</th>
      <th>mean_2</th>
      <th>ratio</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>36</td>
      <td>Ngp</td>
      <td>5.649159e-144</td>
      <td>54.840271</td>
      <td>0.418010</td>
      <td>5.299366</td>
    </tr>
    <tr>
      <th>1</th>
      <td>33</td>
      <td>S100a9</td>
      <td>1.356053e-148</td>
      <td>80.231552</td>
      <td>8.239202</td>
      <td>3.136200</td>
    </tr>
    <tr>
      <th>2</th>
      <td>38</td>
      <td>S100a8</td>
      <td>1.789692e-139</td>
      <td>48.096233</td>
      <td>8.031301</td>
      <td>2.442606</td>
    </tr>
    <tr>
      <th>3</th>
      <td>21</td>
      <td>Elane</td>
      <td>8.662786e-184</td>
      <td>8.812663</td>
      <td>1.012724</td>
      <td>2.285495</td>
    </tr>
    <tr>
      <th>4</th>
      <td>222</td>
      <td>Camp</td>
      <td>2.724041e-46</td>
      <td>4.027065</td>
      <td>0.032162</td>
      <td>2.284047</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2320</th>
      <td>3044</td>
      <td>Tex261</td>
      <td>2.636864e-02</td>
      <td>0.307811</td>
      <td>0.299247</td>
      <td>0.009477</td>
    </tr>
    <tr>
      <th>2321</th>
      <td>3154</td>
      <td>Ndufs4</td>
      <td>3.271385e-02</td>
      <td>0.158712</td>
      <td>0.151988</td>
      <td>0.008397</td>
    </tr>
    <tr>
      <th>2322</th>
      <td>2164</td>
      <td>Bub1b</td>
      <td>1.311236e-03</td>
      <td>0.291533</td>
      <td>0.285507</td>
      <td>0.006747</td>
    </tr>
    <tr>
      <th>2323</th>
      <td>3128</td>
      <td>Socs2</td>
      <td>3.108334e-02</td>
      <td>0.247205</td>
      <td>0.242337</td>
      <td>0.005642</td>
    </tr>
    <tr>
      <th>2324</th>
      <td>3392</td>
      <td>Trpm7</td>
      <td>4.928712e-02</td>
      <td>0.231395</td>
      <td>0.227887</td>
      <td>0.004117</td>
    </tr>
  </tbody>
</table>
<p>2325 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;emb&quot;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Mmp8&#39;</span><span class="p">,</span> <span class="s1">&#39;Ngp&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_48_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_48_0.png" style="width: 1143px; height: 372px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_list</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">dge_gene_A</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">][:</span><span class="mi">20</span><span class="p">])</span><span class="o">+</span><span class="nb">list</span><span class="p">(</span><span class="n">dge_gene_B</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">][:</span><span class="mi">20</span><span class="p">])</span> <span class="c1">#select the top 20 genes from both populations</span>

<span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,[</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">],[</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">]]</span>
<span class="n">renames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neu&#39;</span><span class="p">,</span><span class="s1">&#39;Mon&#39;</span><span class="p">,</span><span class="s1">&#39;Meg-Ery-MBaE&#39;</span><span class="p">,</span><span class="s1">&#39;Lym-Dc&#39;</span><span class="p">]</span>
<span class="n">gene_expression_matrix</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_genes</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span>
                            <span class="n">rename_fates</span><span class="o">=</span><span class="n">renames</span><span class="p">,</span><span class="n">fig_width</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_49_0.png" class="no-scaled-link" src="../../../../_images/docs_doctrees_nbsphinx_day_17_Lineage_Tracing_49_0.png" style="width: 1104px; height: 277px;" />
</div>
</div>
<p>To identify undifferentiated cells that are biased towards either of these lineages, do the following:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">cell_group_A</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;progenitor_transition_map_Neutrophil&#39;</span><span class="p">])</span>
<span class="n">cell_group_B</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;progenitor_transition_map_Monocyte&#39;</span><span class="p">])</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">NameError</span>                                 Traceback (most recent call last)
<span class="ansi-green-fg">/var/folders/24/v4g076gx0mz2mwtd6ktzw8_h0000gn/T/ipykernel_45011/1313473147.py</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-fg">----&gt; 1</span><span class="ansi-red-fg"> </span>cell_group_A<span class="ansi-blue-fg">=</span>np<span class="ansi-blue-fg">.</span>array<span class="ansi-blue-fg">(</span>adata<span class="ansi-blue-fg">.</span>obs<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#39;diff_trajectory_transition_map_Neutrophil&#39;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">      2</span> cell_group_B<span class="ansi-blue-fg">=</span>np<span class="ansi-blue-fg">.</span>array<span class="ansi-blue-fg">(</span>adata<span class="ansi-blue-fg">.</span>obs<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#39;diff_trajectory_transition_map_Monocyte&#39;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">)</span>

<span class="ansi-red-fg">NameError</span>: name &#39;np&#39; is not defined
</pre></div></div>
</div>
<p>Perform differential gene expression analysis between these two groups of undiff cells. You may need to adjust FDR_cutoff.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dge_gene_A</span><span class="p">,</span> <span class="n">dge_gene_B</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">differential_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">cell_group_A</span><span class="o">=</span><span class="n">cell_group_A</span><span class="p">,</span><span class="n">cell_group_B</span><span class="o">=</span><span class="n">cell_group_B</span><span class="p">,</span><span class="n">FDR_cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Patrick Cahan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-LFQLTP4DZG"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-LFQLTP4DZG', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>