{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1dab3056-5dbd-487c-adb3-2d28eb99215b",
   "metadata": {},
   "source": [
    "# Homework #2\n",
    "\n",
    "Due 11:59 pm EST, **Sunday Feb 20th, 2022**\n",
    "Email your solutions (both .ipnb and .html files) to:\n",
    "compscbio@gmail.com.\n",
    "\n",
    "### Goals: force the student to gain practical experience with \n",
    "- annData\n",
    "- scanpy\n",
    "- cleaning scRNAseq data\n",
    "- how parameters impact analysis outcomes\n",
    "- distilling research questions into a set of analytical steps\n",
    "- interpreting results to answer the research questions\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b96582a-33ba-4abd-93ba-2ea7d27b0203",
   "metadata": {},
   "source": [
    "You generated a data set od directed differentation towards neural progenitors. You want to use this data to\n",
    "(1) characterize how cell cycle dynamics change over development\n",
    "(2) determine whether different genes contribute to these differences?\n",
    "\n",
    "To do this, you will need to \n",
    "\n",
    "- clean your data\n",
    "- you need to figure out which day of differentiation each cell came from\n",
    "- count the proportion of cells in each day in each phase\n",
    "- find genes that have expression patterns correlated with S and G2M \n",
    "- propose why/how this might be happening\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e02b1d94-8f08-4e2e-b33b-0eff683e0e59",
   "metadata": {},
   "source": [
    "Prepare this data for the students, then walk through types of analysis that I will ask of them.\n",
    "\n",
    "see Evernote \"rescue negatives from R1 -- take III\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d38e7b25-db88-4682-8538-c0a762848235",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import scipy as sp\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d47b72b4-a154-41e0-8530-5c3a1dd1690a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 7255 × 31065\n",
       "    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'percent_mito'\n",
       "    var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adX1 = sc.read_h5ad(\n",
    "    \"../../../data/cscb/2022/Su_2022/adRaw_QC_012120.h5ad\"\n",
    ")\n",
    "adX1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b7f3f940-02d7-4f53-ae8a-dc721647c487",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Xkr4', 'Gm37381', 'Rp1', 'Sox17', 'Gm37323', 'Mrpl15', 'Rgs20',\n",
       "       'Npbwr1', '4732440D04Rik', 'Gm26901',\n",
       "       ...\n",
       "       'reporter_eGFP', 'reporter_mCherry', 'selection_ampicillin',\n",
       "       'selection_blasticidin-R_pLX304', 'selection_chlormaphenicol',\n",
       "       'selection_hygromycinB', 'selection_kanamycin', 'selection_neomycin',\n",
       "       'selection_puromycin', 'selection_zeocin'],\n",
       "      dtype='object', length=31065)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adX1.var_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "536bcc62-94e7-4c53-82f6-9ca3f8ef2b70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAGAACCAGTAGTTC</th>\n",
       "      <td>5210</td>\n",
       "      <td>8.558527</td>\n",
       "      <td>25133.0</td>\n",
       "      <td>10.131977</td>\n",
       "      <td>21.764214</td>\n",
       "      <td>31.480524</td>\n",
       "      <td>41.165002</td>\n",
       "      <td>55.918514</td>\n",
       "      <td>0.014722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGCTTCTCCCAAC</th>\n",
       "      <td>7130</td>\n",
       "      <td>8.872207</td>\n",
       "      <td>50300.0</td>\n",
       "      <td>10.825780</td>\n",
       "      <td>24.942346</td>\n",
       "      <td>33.980119</td>\n",
       "      <td>43.276342</td>\n",
       "      <td>55.562624</td>\n",
       "      <td>0.006541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACAGGGGTACCCGCA</th>\n",
       "      <td>4647</td>\n",
       "      <td>8.444192</td>\n",
       "      <td>21984.0</td>\n",
       "      <td>9.998116</td>\n",
       "      <td>24.545124</td>\n",
       "      <td>35.530386</td>\n",
       "      <td>45.469432</td>\n",
       "      <td>59.834425</td>\n",
       "      <td>0.015830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGCTAGGAGAGGC</th>\n",
       "      <td>4723</td>\n",
       "      <td>8.460411</td>\n",
       "      <td>22796.0</td>\n",
       "      <td>10.034384</td>\n",
       "      <td>22.148623</td>\n",
       "      <td>31.808212</td>\n",
       "      <td>42.134585</td>\n",
       "      <td>57.865415</td>\n",
       "      <td>0.009344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACACACTCGCTTGCT</th>\n",
       "      <td>3661</td>\n",
       "      <td>8.205765</td>\n",
       "      <td>13217.0</td>\n",
       "      <td>9.489335</td>\n",
       "      <td>27.177120</td>\n",
       "      <td>37.905727</td>\n",
       "      <td>47.461602</td>\n",
       "      <td>61.594916</td>\n",
       "      <td>0.010214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCCTCGTCTTGGTA</th>\n",
       "      <td>3629</td>\n",
       "      <td>8.196988</td>\n",
       "      <td>15046.0</td>\n",
       "      <td>9.618934</td>\n",
       "      <td>29.768709</td>\n",
       "      <td>42.403297</td>\n",
       "      <td>52.113519</td>\n",
       "      <td>65.372857</td>\n",
       "      <td>0.012495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTTCTTGTT</th>\n",
       "      <td>4770</td>\n",
       "      <td>8.470311</td>\n",
       "      <td>24895.0</td>\n",
       "      <td>10.122462</td>\n",
       "      <td>21.807592</td>\n",
       "      <td>32.315726</td>\n",
       "      <td>43.028721</td>\n",
       "      <td>58.859209</td>\n",
       "      <td>0.007351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCGATTCCGATCGG</th>\n",
       "      <td>5657</td>\n",
       "      <td>8.640826</td>\n",
       "      <td>43760.0</td>\n",
       "      <td>10.686499</td>\n",
       "      <td>30.648995</td>\n",
       "      <td>42.746801</td>\n",
       "      <td>52.408592</td>\n",
       "      <td>65.027422</td>\n",
       "      <td>0.016271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCCTCGTTTACCTT</th>\n",
       "      <td>297</td>\n",
       "      <td>5.697093</td>\n",
       "      <td>442.0</td>\n",
       "      <td>6.093570</td>\n",
       "      <td>43.665158</td>\n",
       "      <td>55.429864</td>\n",
       "      <td>78.054299</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.167421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGAGTCATGGATC</th>\n",
       "      <td>6871</td>\n",
       "      <td>8.835210</td>\n",
       "      <td>53622.0</td>\n",
       "      <td>10.889733</td>\n",
       "      <td>22.520607</td>\n",
       "      <td>32.863377</td>\n",
       "      <td>42.956249</td>\n",
       "      <td>57.304838</td>\n",
       "      <td>0.011040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7255 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes_by_counts  log1p_n_genes_by_counts  total_counts  \\\n",
       "AAAGAACCAGTAGTTC               5210                 8.558527       25133.0   \n",
       "AAACGCTTCTCCCAAC               7130                 8.872207       50300.0   \n",
       "AACAGGGGTACCCGCA               4647                 8.444192       21984.0   \n",
       "AAACGCTAGGAGAGGC               4723                 8.460411       22796.0   \n",
       "AACACACTCGCTTGCT               3661                 8.205765       13217.0   \n",
       "...                             ...                      ...           ...   \n",
       "TTTCCTCGTCTTGGTA               3629                 8.196988       15046.0   \n",
       "TTTGGTTGTTCTTGTT               4770                 8.470311       24895.0   \n",
       "TTTCGATTCCGATCGG               5657                 8.640826       43760.0   \n",
       "TTTCCTCGTTTACCTT                297                 5.697093         442.0   \n",
       "TTTGGAGTCATGGATC               6871                 8.835210       53622.0   \n",
       "\n",
       "                  log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "AAAGAACCAGTAGTTC           10.131977                   21.764214   \n",
       "AAACGCTTCTCCCAAC           10.825780                   24.942346   \n",
       "AACAGGGGTACCCGCA            9.998116                   24.545124   \n",
       "AAACGCTAGGAGAGGC           10.034384                   22.148623   \n",
       "AACACACTCGCTTGCT            9.489335                   27.177120   \n",
       "...                              ...                         ...   \n",
       "TTTCCTCGTCTTGGTA            9.618934                   29.768709   \n",
       "TTTGGTTGTTCTTGTT           10.122462                   21.807592   \n",
       "TTTCGATTCCGATCGG           10.686499                   30.648995   \n",
       "TTTCCTCGTTTACCTT            6.093570                   43.665158   \n",
       "TTTGGAGTCATGGATC           10.889733                   22.520607   \n",
       "\n",
       "                  pct_counts_in_top_100_genes  pct_counts_in_top_200_genes  \\\n",
       "AAAGAACCAGTAGTTC                    31.480524                    41.165002   \n",
       "AAACGCTTCTCCCAAC                    33.980119                    43.276342   \n",
       "AACAGGGGTACCCGCA                    35.530386                    45.469432   \n",
       "AAACGCTAGGAGAGGC                    31.808212                    42.134585   \n",
       "AACACACTCGCTTGCT                    37.905727                    47.461602   \n",
       "...                                       ...                          ...   \n",
       "TTTCCTCGTCTTGGTA                    42.403297                    52.113519   \n",
       "TTTGGTTGTTCTTGTT                    32.315726                    43.028721   \n",
       "TTTCGATTCCGATCGG                    42.746801                    52.408592   \n",
       "TTTCCTCGTTTACCTT                    55.429864                    78.054299   \n",
       "TTTGGAGTCATGGATC                    32.863377                    42.956249   \n",
       "\n",
       "                  pct_counts_in_top_500_genes  percent_mito  \n",
       "AAAGAACCAGTAGTTC                    55.918514      0.014722  \n",
       "AAACGCTTCTCCCAAC                    55.562624      0.006541  \n",
       "AACAGGGGTACCCGCA                    59.834425      0.015830  \n",
       "AAACGCTAGGAGAGGC                    57.865415      0.009344  \n",
       "AACACACTCGCTTGCT                    61.594916      0.010214  \n",
       "...                                       ...           ...  \n",
       "TTTCCTCGTCTTGGTA                    65.372857      0.012495  \n",
       "TTTGGTTGTTCTTGTT                    58.859209      0.007351  \n",
       "TTTCGATTCCGATCGG                    65.027422      0.016271  \n",
       "TTTCCTCGTTTACCTT                   100.000000      0.167421  \n",
       "TTTGGAGTCATGGATC                    57.304838      0.011040  \n",
       "\n",
       "[7255 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adX1.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4c44f9c-2f09-46fc-baa7-326bd3ed2f9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AAAGAACCAGTAGTTC', 'AAACGCTTCTCCCAAC', 'AACAGGGGTACCCGCA',\n",
       "       'AAACGCTAGGAGAGGC', 'AACACACTCGCTTGCT', 'AAAGGATAGTCAGAGC',\n",
       "       'AAACGAAGTCCCTAAA', 'AAAGGATCAAGGCTTT', 'AAATGGAGTCCAGTTA',\n",
       "       'AAAGGATCACCCTAAA',\n",
       "       ...\n",
       "       'TTTGTTGCAGTCTTCC', 'TTTCCTCTCGTAGGGA', 'TTTGATCAGCATGATA',\n",
       "       'TTTGGTTTCCCATAAG', 'TTTGTTGCACTCGATA', 'TTTCCTCGTCTTGGTA',\n",
       "       'TTTGGTTGTTCTTGTT', 'TTTCGATTCCGATCGG', 'TTTCCTCGTTTACCTT',\n",
       "       'TTTGGAGTCATGGATC'],\n",
       "      dtype='object', length=7255)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adX1.obs_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "776288f5-cf81-4d5a-8be7-893b0dcf4cff",
   "metadata": {},
   "source": [
    "Get the sample annotation information. See Evernote: \"try to rescue negatives from R1 -- Take II\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9c02654d-22b2-45e2-aded-013894b36918",
   "metadata": {},
   "outputs": [],
   "source": [
    "sBCs = pd.read_csv(\"../../../data/cscb/2022/Su_2022/sampTab_R1_comb_sbarcoded_120519.csv\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0e797016-9bbe-4274-b5f9-1210a0ce3b05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>cell_name</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>sampleClass</th>\n",
       "      <th>treatment</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGGCAGGGA</td>\n",
       "      <td>AAACCCAAGGCAGGGA-1</td>\n",
       "      <td>2566</td>\n",
       "      <td>1228</td>\n",
       "      <td>0.009353</td>\n",
       "      <td>Bar1</td>\n",
       "      <td>pre</td>\n",
       "      <td>0.0</td>\n",
       "      <td>d0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGGTGATCG</td>\n",
       "      <td>AAACCCAAGGTGATCG-1</td>\n",
       "      <td>60138</td>\n",
       "      <td>6821</td>\n",
       "      <td>0.041139</td>\n",
       "      <td>Bar7</td>\n",
       "      <td>WAG</td>\n",
       "      <td>3.0</td>\n",
       "      <td>d3WAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCACACTGAATC</td>\n",
       "      <td>AAACCCACACTGAATC-1</td>\n",
       "      <td>22399</td>\n",
       "      <td>4798</td>\n",
       "      <td>0.047904</td>\n",
       "      <td>Bar9</td>\n",
       "      <td>WAB</td>\n",
       "      <td>4.0</td>\n",
       "      <td>d4WAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCAGTTTACCAG</td>\n",
       "      <td>AAACCCAGTTTACCAG-1</td>\n",
       "      <td>30069</td>\n",
       "      <td>5429</td>\n",
       "      <td>0.051515</td>\n",
       "      <td>Doublet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCATCAATCCAG</td>\n",
       "      <td>AAACCCATCAATCCAG-1</td>\n",
       "      <td>17867</td>\n",
       "      <td>4231</td>\n",
       "      <td>0.072816</td>\n",
       "      <td>Bar9</td>\n",
       "      <td>WAB</td>\n",
       "      <td>4.0</td>\n",
       "      <td>d4WAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7250</th>\n",
       "      <td>TAATCTCCACTACACA</td>\n",
       "      <td>TAATCTCCACTACACA-1</td>\n",
       "      <td>68138</td>\n",
       "      <td>6914</td>\n",
       "      <td>0.040477</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7251</th>\n",
       "      <td>TCCACGTAGGGCAGGA</td>\n",
       "      <td>TCCACGTAGGGCAGGA-1</td>\n",
       "      <td>39383</td>\n",
       "      <td>5814</td>\n",
       "      <td>0.046314</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7252</th>\n",
       "      <td>TTCTCTCCACACCAGC</td>\n",
       "      <td>TTCTCTCCACACCAGC-1</td>\n",
       "      <td>38136</td>\n",
       "      <td>6898</td>\n",
       "      <td>0.043214</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7253</th>\n",
       "      <td>TTGAGTGTCCGCACGA</td>\n",
       "      <td>TTGAGTGTCCGCACGA-1</td>\n",
       "      <td>36697</td>\n",
       "      <td>5678</td>\n",
       "      <td>0.057961</td>\n",
       "      <td>Bar9</td>\n",
       "      <td>WAB</td>\n",
       "      <td>4.0</td>\n",
       "      <td>d4WAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7254</th>\n",
       "      <td>TTGGGATTCAGAACCT</td>\n",
       "      <td>TTGGGATTCAGAACCT-1</td>\n",
       "      <td>33616</td>\n",
       "      <td>5671</td>\n",
       "      <td>0.023858</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7255 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0           cell_name  total_counts  n_genes_by_counts  \\\n",
       "0     AAACCCAAGGCAGGGA  AAACCCAAGGCAGGGA-1          2566               1228   \n",
       "1     AAACCCAAGGTGATCG  AAACCCAAGGTGATCG-1         60138               6821   \n",
       "2     AAACCCACACTGAATC  AAACCCACACTGAATC-1         22399               4798   \n",
       "3     AAACCCAGTTTACCAG  AAACCCAGTTTACCAG-1         30069               5429   \n",
       "4     AAACCCATCAATCCAG  AAACCCATCAATCCAG-1         17867               4231   \n",
       "...                ...                 ...           ...                ...   \n",
       "7250  TAATCTCCACTACACA  TAATCTCCACTACACA-1         68138               6914   \n",
       "7251  TCCACGTAGGGCAGGA  TCCACGTAGGGCAGGA-1         39383               5814   \n",
       "7252  TTCTCTCCACACCAGC  TTCTCTCCACACCAGC-1         38136               6898   \n",
       "7253  TTGAGTGTCCGCACGA  TTGAGTGTCCGCACGA-1         36697               5678   \n",
       "7254  TTGGGATTCAGAACCT  TTGGGATTCAGAACCT-1         33616               5671   \n",
       "\n",
       "      percent_mito sampleClass treatment  timepoint condition  \n",
       "0         0.009353        Bar1       pre        0.0        d0  \n",
       "1         0.041139        Bar7       WAG        3.0     d3WAG  \n",
       "2         0.047904        Bar9       WAB        4.0     d4WAB  \n",
       "3         0.051515     Doublet       NaN        NaN       NaN  \n",
       "4         0.072816        Bar9       WAB        4.0     d4WAB  \n",
       "...            ...         ...       ...        ...       ...  \n",
       "7250      0.040477    Negative       NaN        NaN       NaN  \n",
       "7251      0.046314    Negative       NaN        NaN       NaN  \n",
       "7252      0.043214    Negative       NaN        NaN       NaN  \n",
       "7253      0.057961        Bar9       WAB        4.0     d4WAB  \n",
       "7254      0.023858    Negative       NaN        NaN       NaN  \n",
       "\n",
       "[7255 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sBCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "50bc309a-07b9-4bd4-b261-be4e9c859ced",
   "metadata": {},
   "outputs": [],
   "source": [
    "sBCs = sBCs.rename(columns={\"Unnamed: 0\": \"cell_id\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b78971f0-c1d9-47dc-9178-adbfa839d249",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "adX1 = adX1[ sBCs[\"cell_id\"] ]\n",
    "adX1.obs['timepoint'] = sBCs[\"timepoint\"].values\n",
    "adX1.obs['sampClass'] = sBCs[\"sampleClass\"].values\n",
    "adX1.obs['treatment'] = sBCs[\"treatment\"].values\n",
    "adX1.obs['condition'] = sBCs[\"condition\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f9b1aa90-7468-4a78-8149-cb031b0f3ffd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 7255 × 31065\n",
       "    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'percent_mito', 'timepoint', 'sampClass', 'treatment', 'condition'\n",
       "    var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adX1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a2cf36fd-e923-4ad3-8363-736ebbe2e666",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>sampClass</th>\n",
       "      <th>treatment</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGCAGGGA</th>\n",
       "      <td>1068</td>\n",
       "      <td>6.974479</td>\n",
       "      <td>1996.0</td>\n",
       "      <td>7.599401</td>\n",
       "      <td>27.304609</td>\n",
       "      <td>39.128257</td>\n",
       "      <td>52.204409</td>\n",
       "      <td>71.543086</td>\n",
       "      <td>0.003507</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Bar1</td>\n",
       "      <td>pre</td>\n",
       "      <td>d0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTGATCG</th>\n",
       "      <td>6081</td>\n",
       "      <td>8.713089</td>\n",
       "      <td>43342.0</td>\n",
       "      <td>10.676901</td>\n",
       "      <td>22.241706</td>\n",
       "      <td>32.335841</td>\n",
       "      <td>42.951410</td>\n",
       "      <td>58.077615</td>\n",
       "      <td>0.012644</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Bar7</td>\n",
       "      <td>WAG</td>\n",
       "      <td>d3WAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACACTGAATC</th>\n",
       "      <td>4204</td>\n",
       "      <td>8.344030</td>\n",
       "      <td>16212.0</td>\n",
       "      <td>9.693568</td>\n",
       "      <td>20.207254</td>\n",
       "      <td>30.039477</td>\n",
       "      <td>40.260301</td>\n",
       "      <td>56.032568</td>\n",
       "      <td>0.015421</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Bar9</td>\n",
       "      <td>WAB</td>\n",
       "      <td>d4WAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTTTACCAG</th>\n",
       "      <td>4789</td>\n",
       "      <td>8.474286</td>\n",
       "      <td>21514.0</td>\n",
       "      <td>9.976505</td>\n",
       "      <td>21.023520</td>\n",
       "      <td>30.598680</td>\n",
       "      <td>40.438784</td>\n",
       "      <td>55.735800</td>\n",
       "      <td>0.011946</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Doublet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCATCAATCCAG</th>\n",
       "      <td>3703</td>\n",
       "      <td>8.217169</td>\n",
       "      <td>12508.0</td>\n",
       "      <td>9.434204</td>\n",
       "      <td>24.112568</td>\n",
       "      <td>35.297410</td>\n",
       "      <td>45.738727</td>\n",
       "      <td>60.233451</td>\n",
       "      <td>0.015750</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Bar9</td>\n",
       "      <td>WAB</td>\n",
       "      <td>d4WAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAATCTCCACTACACA</th>\n",
       "      <td>6187</td>\n",
       "      <td>8.730367</td>\n",
       "      <td>48935.0</td>\n",
       "      <td>10.798268</td>\n",
       "      <td>21.491775</td>\n",
       "      <td>31.590886</td>\n",
       "      <td>41.910698</td>\n",
       "      <td>57.132931</td>\n",
       "      <td>0.011485</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCACGTAGGGCAGGA</th>\n",
       "      <td>5136</td>\n",
       "      <td>8.544225</td>\n",
       "      <td>28158.0</td>\n",
       "      <td>10.245623</td>\n",
       "      <td>23.204773</td>\n",
       "      <td>34.160807</td>\n",
       "      <td>44.204134</td>\n",
       "      <td>59.070246</td>\n",
       "      <td>0.012892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCTCTCCACACCAGC</th>\n",
       "      <td>6130</td>\n",
       "      <td>8.721113</td>\n",
       "      <td>28591.0</td>\n",
       "      <td>10.260882</td>\n",
       "      <td>20.314085</td>\n",
       "      <td>28.725823</td>\n",
       "      <td>38.099402</td>\n",
       "      <td>51.547690</td>\n",
       "      <td>0.008989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGAGTGTCCGCACGA</th>\n",
       "      <td>5002</td>\n",
       "      <td>8.517793</td>\n",
       "      <td>25894.0</td>\n",
       "      <td>10.161805</td>\n",
       "      <td>26.832471</td>\n",
       "      <td>38.414304</td>\n",
       "      <td>48.042017</td>\n",
       "      <td>61.489148</td>\n",
       "      <td>0.017958</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Bar9</td>\n",
       "      <td>WAB</td>\n",
       "      <td>d4WAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGGATTCAGAACCT</th>\n",
       "      <td>5004</td>\n",
       "      <td>8.518193</td>\n",
       "      <td>24051.0</td>\n",
       "      <td>10.087974</td>\n",
       "      <td>24.514573</td>\n",
       "      <td>35.038044</td>\n",
       "      <td>44.605214</td>\n",
       "      <td>58.800050</td>\n",
       "      <td>0.006902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7255 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes_by_counts  log1p_n_genes_by_counts  total_counts  \\\n",
       "AAACCCAAGGCAGGGA               1068                 6.974479        1996.0   \n",
       "AAACCCAAGGTGATCG               6081                 8.713089       43342.0   \n",
       "AAACCCACACTGAATC               4204                 8.344030       16212.0   \n",
       "AAACCCAGTTTACCAG               4789                 8.474286       21514.0   \n",
       "AAACCCATCAATCCAG               3703                 8.217169       12508.0   \n",
       "...                             ...                      ...           ...   \n",
       "TAATCTCCACTACACA               6187                 8.730367       48935.0   \n",
       "TCCACGTAGGGCAGGA               5136                 8.544225       28158.0   \n",
       "TTCTCTCCACACCAGC               6130                 8.721113       28591.0   \n",
       "TTGAGTGTCCGCACGA               5002                 8.517793       25894.0   \n",
       "TTGGGATTCAGAACCT               5004                 8.518193       24051.0   \n",
       "\n",
       "                  log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "AAACCCAAGGCAGGGA            7.599401                   27.304609   \n",
       "AAACCCAAGGTGATCG           10.676901                   22.241706   \n",
       "AAACCCACACTGAATC            9.693568                   20.207254   \n",
       "AAACCCAGTTTACCAG            9.976505                   21.023520   \n",
       "AAACCCATCAATCCAG            9.434204                   24.112568   \n",
       "...                              ...                         ...   \n",
       "TAATCTCCACTACACA           10.798268                   21.491775   \n",
       "TCCACGTAGGGCAGGA           10.245623                   23.204773   \n",
       "TTCTCTCCACACCAGC           10.260882                   20.314085   \n",
       "TTGAGTGTCCGCACGA           10.161805                   26.832471   \n",
       "TTGGGATTCAGAACCT           10.087974                   24.514573   \n",
       "\n",
       "                  pct_counts_in_top_100_genes  pct_counts_in_top_200_genes  \\\n",
       "AAACCCAAGGCAGGGA                    39.128257                    52.204409   \n",
       "AAACCCAAGGTGATCG                    32.335841                    42.951410   \n",
       "AAACCCACACTGAATC                    30.039477                    40.260301   \n",
       "AAACCCAGTTTACCAG                    30.598680                    40.438784   \n",
       "AAACCCATCAATCCAG                    35.297410                    45.738727   \n",
       "...                                       ...                          ...   \n",
       "TAATCTCCACTACACA                    31.590886                    41.910698   \n",
       "TCCACGTAGGGCAGGA                    34.160807                    44.204134   \n",
       "TTCTCTCCACACCAGC                    28.725823                    38.099402   \n",
       "TTGAGTGTCCGCACGA                    38.414304                    48.042017   \n",
       "TTGGGATTCAGAACCT                    35.038044                    44.605214   \n",
       "\n",
       "                  pct_counts_in_top_500_genes  percent_mito  timepoint  \\\n",
       "AAACCCAAGGCAGGGA                    71.543086      0.003507        0.0   \n",
       "AAACCCAAGGTGATCG                    58.077615      0.012644        3.0   \n",
       "AAACCCACACTGAATC                    56.032568      0.015421        4.0   \n",
       "AAACCCAGTTTACCAG                    55.735800      0.011946        NaN   \n",
       "AAACCCATCAATCCAG                    60.233451      0.015750        4.0   \n",
       "...                                       ...           ...        ...   \n",
       "TAATCTCCACTACACA                    57.132931      0.011485        NaN   \n",
       "TCCACGTAGGGCAGGA                    59.070246      0.012892        NaN   \n",
       "TTCTCTCCACACCAGC                    51.547690      0.008989        NaN   \n",
       "TTGAGTGTCCGCACGA                    61.489148      0.017958        4.0   \n",
       "TTGGGATTCAGAACCT                    58.800050      0.006902        NaN   \n",
       "\n",
       "                 sampClass treatment condition  \n",
       "AAACCCAAGGCAGGGA      Bar1       pre        d0  \n",
       "AAACCCAAGGTGATCG      Bar7       WAG     d3WAG  \n",
       "AAACCCACACTGAATC      Bar9       WAB     d4WAB  \n",
       "AAACCCAGTTTACCAG   Doublet       NaN       NaN  \n",
       "AAACCCATCAATCCAG      Bar9       WAB     d4WAB  \n",
       "...                    ...       ...       ...  \n",
       "TAATCTCCACTACACA  Negative       NaN       NaN  \n",
       "TCCACGTAGGGCAGGA  Negative       NaN       NaN  \n",
       "TTCTCTCCACACCAGC  Negative       NaN       NaN  \n",
       "TTGAGTGTCCGCACGA      Bar9       WAB     d4WAB  \n",
       "TTGGGATTCAGAACCT  Negative       NaN       NaN  \n",
       "\n",
       "[7255 rows x 13 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adX1.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "71039f20-d91d-4cfe-8efe-019cd9fc5c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "rcBCs = pd.read_csv(\"../../../data/cscb/2022/Su_2022/sampTab_R1_SBC_rescuedv2_012020.csv\", header=0)\n",
    "rcBCs = rcBCs.rename(columns={\"Unnamed: 0\": \"cell_id\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "004a09f6-a326-48e1-874a-d61751900b78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>cell_name</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>treatment</th>\n",
       "      <th>condition</th>\n",
       "      <th>sampClass</th>\n",
       "      <th>leiden</th>\n",
       "      <th>...</th>\n",
       "      <th>Bar5</th>\n",
       "      <th>Bar6</th>\n",
       "      <th>Bar7</th>\n",
       "      <th>Bar8</th>\n",
       "      <th>Bar9</th>\n",
       "      <th>Bar10</th>\n",
       "      <th>Bar11</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nUMI_total</th>\n",
       "      <th>newClass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGGTGATCG</td>\n",
       "      <td>AAACCCAAGGTGATCG</td>\n",
       "      <td>43342</td>\n",
       "      <td>6081</td>\n",
       "      <td>0.012644</td>\n",
       "      <td>3.0</td>\n",
       "      <td>WAG</td>\n",
       "      <td>d3WAG</td>\n",
       "      <td>Bar7</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>103</td>\n",
       "      <td>Bar7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGTCACAGG</td>\n",
       "      <td>AAACCCAAGTCACAGG</td>\n",
       "      <td>34176</td>\n",
       "      <td>5464</td>\n",
       "      <td>0.017878</td>\n",
       "      <td>2.0</td>\n",
       "      <td>pre</td>\n",
       "      <td>d2</td>\n",
       "      <td>Negative</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>99</td>\n",
       "      <td>109</td>\n",
       "      <td>Bar3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTGCAGGT</td>\n",
       "      <td>AAACCCAAGTGCAGGT</td>\n",
       "      <td>48456</td>\n",
       "      <td>6537</td>\n",
       "      <td>0.008461</td>\n",
       "      <td>3.0</td>\n",
       "      <td>WA</td>\n",
       "      <td>d3WA</td>\n",
       "      <td>Negative</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>8</td>\n",
       "      <td>143</td>\n",
       "      <td>153</td>\n",
       "      <td>Bar4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCACACTGAATC</td>\n",
       "      <td>AAACCCACACTGAATC</td>\n",
       "      <td>16212</td>\n",
       "      <td>4204</td>\n",
       "      <td>0.015421</td>\n",
       "      <td>4.0</td>\n",
       "      <td>WAB</td>\n",
       "      <td>d4WAB</td>\n",
       "      <td>Bar9</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>187</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>254</td>\n",
       "      <td>267</td>\n",
       "      <td>Bar9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCACATCGGTTA</td>\n",
       "      <td>AAACCCACATCGGTTA</td>\n",
       "      <td>41881</td>\n",
       "      <td>5855</td>\n",
       "      <td>0.019914</td>\n",
       "      <td>0.0</td>\n",
       "      <td>pre</td>\n",
       "      <td>d0</td>\n",
       "      <td>Negative</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>86</td>\n",
       "      <td>93</td>\n",
       "      <td>Bar1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6061</th>\n",
       "      <td>TTTGTTGCATCGGCCA</td>\n",
       "      <td>TTTGTTGCATCGGCCA</td>\n",
       "      <td>25648</td>\n",
       "      <td>5155</td>\n",
       "      <td>0.012516</td>\n",
       "      <td>1.0</td>\n",
       "      <td>pre</td>\n",
       "      <td>d1</td>\n",
       "      <td>Negative</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>95</td>\n",
       "      <td>101</td>\n",
       "      <td>Bar2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6062</th>\n",
       "      <td>TTTGTTGCATGAGATA</td>\n",
       "      <td>TTTGTTGCATGAGATA</td>\n",
       "      <td>17995</td>\n",
       "      <td>4216</td>\n",
       "      <td>0.013226</td>\n",
       "      <td>3.0</td>\n",
       "      <td>WAN</td>\n",
       "      <td>d3WAN</td>\n",
       "      <td>Bar6</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>84</td>\n",
       "      <td>90</td>\n",
       "      <td>Bar6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6063</th>\n",
       "      <td>TTTGTTGTCCAGCTCT</td>\n",
       "      <td>TTTGTTGTCCAGCTCT</td>\n",
       "      <td>28854</td>\n",
       "      <td>5477</td>\n",
       "      <td>0.012581</td>\n",
       "      <td>4.0</td>\n",
       "      <td>WA</td>\n",
       "      <td>d4WA</td>\n",
       "      <td>Bar8</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>114</td>\n",
       "      <td>118</td>\n",
       "      <td>Bar8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6064</th>\n",
       "      <td>TTTGTTGTCCTACGAA</td>\n",
       "      <td>TTTGTTGTCCTACGAA</td>\n",
       "      <td>31228</td>\n",
       "      <td>4286</td>\n",
       "      <td>0.012233</td>\n",
       "      <td>1.0</td>\n",
       "      <td>pre</td>\n",
       "      <td>d1</td>\n",
       "      <td>Bar2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>1476</td>\n",
       "      <td>1656</td>\n",
       "      <td>Bar2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6065</th>\n",
       "      <td>TTTGTTGTCTTACTGT</td>\n",
       "      <td>TTTGTTGTCTTACTGT</td>\n",
       "      <td>33263</td>\n",
       "      <td>5962</td>\n",
       "      <td>0.010913</td>\n",
       "      <td>3.0</td>\n",
       "      <td>WAG</td>\n",
       "      <td>d3WAG</td>\n",
       "      <td>Bar7</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>706</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>793</td>\n",
       "      <td>844</td>\n",
       "      <td>Bar7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6066 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               cell_id         cell_name  total_counts  n_genes_by_counts  \\\n",
       "0     AAACCCAAGGTGATCG  AAACCCAAGGTGATCG         43342               6081   \n",
       "1     AAACCCAAGTCACAGG  AAACCCAAGTCACAGG         34176               5464   \n",
       "2     AAACCCAAGTGCAGGT  AAACCCAAGTGCAGGT         48456               6537   \n",
       "3     AAACCCACACTGAATC  AAACCCACACTGAATC         16212               4204   \n",
       "4     AAACCCACATCGGTTA  AAACCCACATCGGTTA         41881               5855   \n",
       "...                ...               ...           ...                ...   \n",
       "6061  TTTGTTGCATCGGCCA  TTTGTTGCATCGGCCA         25648               5155   \n",
       "6062  TTTGTTGCATGAGATA  TTTGTTGCATGAGATA         17995               4216   \n",
       "6063  TTTGTTGTCCAGCTCT  TTTGTTGTCCAGCTCT         28854               5477   \n",
       "6064  TTTGTTGTCCTACGAA  TTTGTTGTCCTACGAA         31228               4286   \n",
       "6065  TTTGTTGTCTTACTGT  TTTGTTGTCTTACTGT         33263               5962   \n",
       "\n",
       "      percent_mito  timepoint treatment condition sampClass  leiden  ...  \\\n",
       "0         0.012644        3.0       WAG     d3WAG      Bar7       0  ...   \n",
       "1         0.017878        2.0       pre        d2  Negative       0  ...   \n",
       "2         0.008461        3.0        WA      d3WA  Negative       0  ...   \n",
       "3         0.015421        4.0       WAB     d4WAB      Bar9       1  ...   \n",
       "4         0.019914        0.0       pre        d0  Negative       2  ...   \n",
       "...            ...        ...       ...       ...       ...     ...  ...   \n",
       "6061      0.012516        1.0       pre        d1  Negative       2  ...   \n",
       "6062      0.013226        3.0       WAN     d3WAN      Bar6       0  ...   \n",
       "6063      0.012581        4.0        WA      d4WA      Bar8       0  ...   \n",
       "6064      0.012233        1.0       pre        d1      Bar2       0  ...   \n",
       "6065      0.010913        3.0       WAG     d3WAG      Bar7       1  ...   \n",
       "\n",
       "      Bar5  Bar6  Bar7  Bar8  Bar9  Bar10  Bar11  nUMI  nUMI_total  newClass  \n",
       "0        2     1    26     5     5      8      6    95         103      Bar7  \n",
       "1        1     2     2     1     6      7      6    99         109      Bar3  \n",
       "2        4     4     0     3     4     19      8   143         153      Bar4  \n",
       "3        3     2     1     4   187      4     11   254         267      Bar9  \n",
       "4        5     3     1     2     3      8      5    86          93      Bar1  \n",
       "...    ...   ...   ...   ...   ...    ...    ...   ...         ...       ...  \n",
       "6061     0     1     3     7     9      5      9    95         101      Bar2  \n",
       "6062     1    21    16     1     9      3      7    84          90      Bar6  \n",
       "6063     0     3     1    59    25      2      5   114         118      Bar8  \n",
       "6064     0     3     1     3     5      7      4  1476        1656      Bar2  \n",
       "6065     6     5   706     5     9     12      4   793         844      Bar7  \n",
       "\n",
       "[6066 rows x 24 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del(sBCs2)\n",
    "rcBCs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec8622ec-d94d-4ff5-9e84-8a99737f5b52",
   "metadata": {},
   "source": [
    "Make another annData with this filtered/rescued data\n",
    "\n",
    "Then figure out what is unique to the first data set, and make an annData from that\n",
    "\n",
    "Then combine the two\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da56804-7d43-489a-a1e9-1fb226340ce2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7641d7b5-f680-4995-b241-bbd457c07eff",
   "metadata": {},
   "source": [
    "Do basic analysis pipeline to identify\n",
    "\n",
    "- low quality cells\n",
    "- potential doublets\n",
    "- cells off of the esc -> neural progenitor trajectory\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
