<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>scRNAseq analysis &mdash; cscb2022  documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="scRNAseq analysis part 2" href="07_scRNAseq_analysis_part_2.html" />
    <link rel="prev" title="scRNAseq data generation" href="05_scRNAseq_data_generation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> cscb2022<img src="../_static/CSCB_logo2.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Schedule</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../schedule.html">Schedule overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../schedule.html#homework-and-final-project-due-dates">Homework and final project due dates</a></li>
<li class="toctree-l1"><a class="reference internal" href="../schedule.html#lecture-schedule">Lecture schedule</a></li>
</ul>
<p class="caption"><span class="caption-text">Lectures:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="01_history_of_stem_cells.html">History of Stem Cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="02_introduction_to_python_part_1.html">Comp Lesson 1: Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="03_introduction_to_python_part_2.html">Comp Lesson 2: Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="04_Till_McCulloch.html">Till and McCulloch</a></li>
<li class="toctree-l1"><a class="reference internal" href="05_scRNAseq_data_generation.html">scRNAseq data generation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">scRNAseq analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#UMI-counting:">UMI counting:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Detecting-cells">Detecting cells</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#CellRanger-output-status-(web-summary)">CellRanger output status (web summary)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Sequencing-saturation:">Sequencing saturation:</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Post-UMI-counting-analysis">Post UMI counting analysis</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="07_scRNAseq_analysis_part_2.html">scRNAseq analysis part 2</a></li>
<li class="toctree-l1"><a class="reference internal" href="08_cell_identity.html">Cell identity</a></li>
<li class="toctree-l1"><a class="reference internal" href="09_trajectory_inference.html">Trajectory inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="10_TI_part_2.html">How to do TI analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="11_stemness.html">How to do ‘Stemness’ analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="12_RNA_velocity.html">RNA Velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="13_scvelo.html">RNA velocity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="14_Lineage_tracing.html">Lineage tracing</a></li>
<li class="toctree-l1"><a class="reference internal" href="15_Pluripotency.html">Pluripotency</a></li>
</ul>
<p class="caption"><span class="caption-text">Homework</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../homework/HW_01.html">Homework #1</a></li>
<li class="toctree-l1"><a class="reference internal" href="../homework/HW_02.html">Homework #2</a></li>
<li class="toctree-l1"><a class="reference internal" href="../homework/HW_03.html">Homework #3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../homework/HW_04.html">Homework #4</a></li>
<li class="toctree-l1"><a class="reference internal" href="../homework/HW_05.html">Homework #5</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cscb2022</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>scRNAseq analysis</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/day/06_scRNAseq_analysis.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="scRNAseq-analysis">
<h1>scRNAseq analysis<a class="headerlink" href="#scRNAseq-analysis" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger</a></p></li>
<li><p>cellranger <code class="docutils literal notranslate"><span class="pre">count</span></code>: Counts cell barcodes, aligns to genome/transcriptome, counts UMIs</p></li>
</ul>
<p>Example invocation:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$ cellranger count --id=sample345 \
                   --transcriptome=/opt/refdata-gex-GRCh38-2020-A \
                   --fastqs=/home/jdoe/runs/HAWT7ADXX/outs/fastq_path \
                   --sample=mysample \
                   --expect-cells=1000 \
                   --localcores=8 \
                   --localmem=64
</pre></div>
</div>
<p>Example output:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Outputs:
- Run summary HTML:                         /opt/sample345/outs/web_summary.html
- Run summary CSV:                          /opt/sample345/outs/metrics_summary.csv
- BAM:                                      /opt/sample345/outs/possorted_genome_bam.bam
- BAM index:                                /opt/sample345/outs/possorted_genome_bam.bam.bai
- Filtered feature-barcode matrices MEX:    /opt/sample345/outs/filtered_feature_bc_matrix
- Filtered feature-barcode matrices HDF5:   /opt/sample345/outs/filtered_feature_bc_matrix.h5
- Unfiltered feature-barcode matrices MEX:  /opt/sample345/outs/raw_feature_bc_matrix
- Unfiltered feature-barcode matrices HDF5: /opt/sample345/outs/raw_feature_bc_matrix.h5
- Secondary analysis output CSV:            /opt/sample345/outs/analysis
- Per-molecule read information:            /opt/sample345/outs/molecule_info.h5
- CRISPR-specific analysis:                 null
- Loupe Browser file:                       /opt/sample345/outs/cloupe.cloupe
- Feature Reference:                        null
- Target Panel File:                        null

Waiting 6 seconds for UI to do final refresh.
Pipestance completed successfully!

yyyy-mm-dd hh:mm:ss Shutting down.
Saving pipestance info to &quot;tiny/tiny.mri.tgz&quot;
</pre></div>
</div>
<p>What is actually doing?</p>
<ul class="simple">
<li><p>Trims reads</p></li>
<li><p>Splicing aware alignment to reference genome</p></li>
<li><p>GTF file to assign aligned reads to transcripts (see below)</p></li>
<li><p>handles reads mapping to &gt;1 locus</p></li>
</ul>
<p>How does it handle reads that hit different parts of the genome?</p>
<p><img alt="4e68a324bb8844398f6ce03f162bb73b" src="../_images/gex-processing-introns.png" /></p>
<ul class="simple">
<li><p>aligned reads for UMI counting, by default, are ’ transcriptomic’ (blue) reads</p></li>
<li><p>Often useful an more appropriate to include exonic and intronic aligned reads, too</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">include-introns</span></code>: use all but antisense reads for UMI counting</p></li>
</ul>
<div class="section" id="UMI-counting:">
<h2>UMI counting:<a class="headerlink" href="#UMI-counting:" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>group reads with same cell bc, umi, gene</p></li>
<li><p>correct umis (one base off)</p></li>
<li><p>cell bc, umi, diff gene -&gt; only keep gene with most support</p></li>
<li><p>discard both if tied</p></li>
<li><p>all reads with same cell bc, umi, gene are counted as one UMI</p></li>
<li><p>number of reads contributing to that UMI are stored in molecule info file (useful later)</p></li>
</ul>
</div>
<div class="section" id="Detecting-cells">
<h2>Detecting cells<a class="headerlink" href="#Detecting-cells" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>EmptyDrops method (Lun et al 2018)</p></li>
<li><p>ID true positives with total UMI threshold determine based on total UMIs in top fraction of expected numvber of barcodes</p></li>
<li><p>Select set of true negatives and compare complexity of remaining barcodes to it</p></li>
</ul>
<p><img alt="530d2fce988c4153a5d8d7516d552eac" src="../_images/knee-plot-old-cell-calling.png" /></p>
<p><img alt="3f9f8c8fbcb041c483b2335bedc7b7ac" src="../_images/knee-plot-new-cell-calling.png" /></p>
<div class="section" id="CellRanger-output-status-(web-summary)">
<h3>CellRanger output status (web summary)<a class="headerlink" href="#CellRanger-output-status-(web-summary)" title="Permalink to this headline">¶</a></h3>
<p><img alt="7334b721f2944b758da52feefbe02e23" src="../_images/web-summary-gex-3.1a.png" /></p>
</div>
</div>
<div class="section" id="Sequencing-saturation:">
<h2>Sequencing saturation:<a class="headerlink" href="#Sequencing-saturation:" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>“If sequencing saturation is at 50%, it means that every 2 new reads will result in 1 new UMI count (unique transcript) detected.”</p></li>
<li><p>Calcualted based on the ratio of unique to duplicate UMIs</p></li>
</ul>
<hr><div class="section" id="Post-UMI-counting-analysis">
<h3>Post UMI counting analysis<a class="headerlink" href="#Post-UMI-counting-analysis" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Loading data</p></li>
<li><p>Standard quality control</p></li>
<li><p>Some fun, downstream stuff</p></li>
</ul>
<p>Scanpy <a class="reference external" href="https://scanpy.readthedocs.io/en/stable">https://scanpy.readthedocs.io/en/stable</a>/# is the package that will be the workhorse for much of what we do in the class,</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">pip</span> install scanpy
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scipy</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
</pre></div>
</div>
</div>
<p>You can download the data used below from here. It is a zip file containing the market matrix files that I showed you in class: genes.tsv, matrix.mtx, and barcodes.tsv. You will need to unzip it, and then change the path argument below so that it is pointing to wherever you unzipped the 3 files.</p>
<p><a class="reference external" href="https://compstemcellbio.s3.amazonaws.com/2022/data/d4_scRNAseq_data.zip">link to data if not appearing above</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_mtx</span><span class="p">(</span><span class="s2">&quot;data/d4/&quot;</span><span class="p">,</span> <span class="n">var_names</span><span class="o">=</span><span class="s1">&#39;gene_symbols&#39;</span><span class="p">,</span><span class="n">cache</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">gex_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 5405 × 27998
    var: &#39;gene_ids&#39;
</pre></div></div>
</div>
<p>What is an anndata?</p>
<p><img alt="c29a734fba90496aa6ecfdb62f46fe6f" src="../_images/anndata.png" /></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACATACCCTACC-1</th>
    </tr>
    <tr>
      <th>AAACATACGTCGTA-1</th>
    </tr>
    <tr>
      <th>AAACATACTTTCAC-1</th>
    </tr>
    <tr>
      <th>AAACATTGCATTGG-1</th>
    </tr>
    <tr>
      <th>AAACATTGCTTGCC-1</th>
    </tr>
    <tr>
      <th>...</th>
    </tr>
    <tr>
      <th>TTTGACTGAGGCGA-1</th>
    </tr>
    <tr>
      <th>TTTGACTGCATTGG-1</th>
    </tr>
    <tr>
      <th>TTTGACTGCTGGAT-1</th>
    </tr>
    <tr>
      <th>TTTGACTGGTGAGG-1</th>
    </tr>
    <tr>
      <th>TTTGACTGTACAGC-1</th>
    </tr>
  </tbody>
</table>
<p>5405 rows × 0 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;AAACATACCCTACC-1&#39;, &#39;AAACATACGTCGTA-1&#39;, &#39;AAACATACTTTCAC-1&#39;,
       &#39;AAACATTGCATTGG-1&#39;, &#39;AAACATTGCTTGCC-1&#39;, &#39;AAACATTGGACGAG-1&#39;,
       &#39;AAACATTGGGTCTA-1&#39;, &#39;AAACATTGTGACCA-1&#39;, &#39;AAACCGTGAAAGCA-1&#39;,
       &#39;AAACCGTGAACGTC-1&#39;,
       ...
       &#39;TTTCTACTGACAAA-1&#39;, &#39;TTTCTACTGCGGAA-1&#39;, &#39;TTTCTACTTGCATG-1&#39;,
       &#39;TTTCTACTTGGTAC-1&#39;, &#39;TTTGACTGACTCTT-1&#39;, &#39;TTTGACTGAGGCGA-1&#39;,
       &#39;TTTGACTGCATTGG-1&#39;, &#39;TTTGACTGCTGGAT-1&#39;, &#39;TTTGACTGGTGAGG-1&#39;,
       &#39;TTTGACTGTACAGC-1&#39;],
      dtype=&#39;object&#39;, length=5405)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_ids</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Xkr4</th>
      <td>ENSMUSG00000051951</td>
    </tr>
    <tr>
      <th>Gm1992</th>
      <td>ENSMUSG00000089699</td>
    </tr>
    <tr>
      <th>Gm37381</th>
      <td>ENSMUSG00000102343</td>
    </tr>
    <tr>
      <th>Rp1</th>
      <td>ENSMUSG00000025900</td>
    </tr>
    <tr>
      <th>Rp1-1</th>
      <td>ENSMUSG00000109048</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>AC168977.1</th>
      <td>ENSMUSG00000079808</td>
    </tr>
    <tr>
      <th>PISD</th>
      <td>ENSMUSG00000095041</td>
    </tr>
    <tr>
      <th>DHRSX</th>
      <td>ENSMUSG00000063897</td>
    </tr>
    <tr>
      <th>Vmn2r122</th>
      <td>ENSMUSG00000096730</td>
    </tr>
    <tr>
      <th>CAAA01147332.1</th>
      <td>ENSMUSG00000095742</td>
    </tr>
  </tbody>
</table>
<p>27998 rows × 1 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;Xkr4&#39;, &#39;Gm1992&#39;, &#39;Gm37381&#39;, &#39;Rp1&#39;, &#39;Rp1-1&#39;, &#39;Sox17&#39;, &#39;Gm37323&#39;,
       &#39;Mrpl15&#39;, &#39;Lypla1&#39;, &#39;Gm37988&#39;,
       ...
       &#39;AC125149.1&#39;, &#39;AC125149.2&#39;, &#39;AC125149.4&#39;, &#39;AC234645.1&#39;, &#39;AC168977.2&#39;,
       &#39;AC168977.1&#39;, &#39;PISD&#39;, &#39;DHRSX&#39;, &#39;Vmn2r122&#39;, &#39;CAAA01147332.1&#39;],
      dtype=&#39;object&#39;, length=27998)
</pre></div></div>
</div>
<p>We can add to obs. For example, I want to add some sample annotation</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;sampleName&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;mEB_day4&quot;</span>
</pre></div>
</div>
</div>
<p>Let’s look at how the plotting functions work.</p>
<p>Note that, in general, scanpy has 3 classes of functions:</p>
<p>sc.pp: pre-processing functions sc.tl: tools sc.pl: plotting</p>
<p>In the example below, the function <code class="docutils literal notranslate"><span class="pre">highest_expr_genes</span></code> identifies the n_top genes with highest mean expression, and then passes the expression of these on to sns.boxplot function:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/day_06_scRNAseq_analysis_27_0.png" src="../_images/day_06_scRNAseq_analysis_27_0.png" />
</div>
</div>
<p>You can use the **kwrds feature to pass arguments directly through to sns.boxplot as name=value pairs. Here, we set palette and width args:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=.</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/day_06_scRNAseq_analysis_29_0.png" src="../_images/day_06_scRNAseq_analysis_29_0.png" />
</div>
</div>
<p>We can alter some of the default plotting parameters, too, like the resolution</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=.</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/day_06_scRNAseq_analysis_32_0.png" class="no-scaled-link" src="../_images/day_06_scRNAseq_analysis_32_0.png" style="width: 406px; height: 391px;" />
</div>
</div>
<p>It is commonplace to calculate the percent of reads that come from mitocondrially-encoded genes and ribosomal genes. ** NB ** In lecture I said that the ribosomal genes shown above are not poly-adenylated. This was incorrect. These transcripts <strong>are</strong> polyadenylated. Thank you to the student who pointed this out after class!</p>
<p>The reasons that we like to calculate these values is as follows:</p>
<ul class="simple">
<li><p>mt genes are thought to be relatively more persistent in dying or damaged cells.</p></li>
<li><p>ribosomal genes are highly abundant. Thus barcodes in which ribosomal genes make up the vast proportion of UMIs might yield low sensitivity for all other transcripts.</p></li>
</ul>
<p>We can readily find both classes genes based on their symbols. Let’s add this to .var:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;mt&#39;</span><span class="p">]</span><span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">((</span><span class="s2">&quot;mt-&quot;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;mt&#39;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
13
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;ribo&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">((</span><span class="s2">&quot;Rps&quot;</span><span class="p">,</span><span class="s2">&quot;Rpl&quot;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;ribo&#39;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
113
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_ids</th>
      <th>mt</th>
      <th>ribo</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Xkr4</th>
      <td>ENSMUSG00000051951</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Gm1992</th>
      <td>ENSMUSG00000089699</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Gm37381</th>
      <td>ENSMUSG00000102343</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Rp1</th>
      <td>ENSMUSG00000025900</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Rp1-1</th>
      <td>ENSMUSG00000109048</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>AC168977.1</th>
      <td>ENSMUSG00000079808</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>PISD</th>
      <td>ENSMUSG00000095041</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>DHRSX</th>
      <td>ENSMUSG00000063897</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Vmn2r122</th>
      <td>ENSMUSG00000096730</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>CAAA01147332.1</th>
      <td>ENSMUSG00000095742</td>
      <td>False</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>27998 rows × 3 columns</p>
</div></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">sc.pp.calculate_qc_metrics</span></code> calculates the specified quantities (and several others by default) and adds them to the .obs data frame</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">qc_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ribo&#39;</span><span class="p">,</span> <span class="s1">&#39;mt&#39;</span><span class="p">],</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sampleName</th>
      <th>n_genes_by_counts</th>
      <th>total_counts</th>
      <th>total_counts_ribo</th>
      <th>pct_counts_ribo</th>
      <th>total_counts_mt</th>
      <th>pct_counts_mt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACATACCCTACC-1</th>
      <td>mEB_day4</td>
      <td>1212</td>
      <td>2238.0</td>
      <td>629.0</td>
      <td>28.105453</td>
      <td>28.0</td>
      <td>1.251117</td>
    </tr>
    <tr>
      <th>AAACATACGTCGTA-1</th>
      <td>mEB_day4</td>
      <td>1588</td>
      <td>3831.0</td>
      <td>1267.0</td>
      <td>33.072304</td>
      <td>34.0</td>
      <td>0.887497</td>
    </tr>
    <tr>
      <th>AAACATACTTTCAC-1</th>
      <td>mEB_day4</td>
      <td>1538</td>
      <td>3381.0</td>
      <td>961.0</td>
      <td>28.423544</td>
      <td>2.0</td>
      <td>0.059154</td>
    </tr>
    <tr>
      <th>AAACATTGCATTGG-1</th>
      <td>mEB_day4</td>
      <td>1221</td>
      <td>2489.0</td>
      <td>750.0</td>
      <td>30.132584</td>
      <td>24.0</td>
      <td>0.964243</td>
    </tr>
    <tr>
      <th>AAACATTGCTTGCC-1</th>
      <td>mEB_day4</td>
      <td>2661</td>
      <td>9510.0</td>
      <td>3132.0</td>
      <td>32.933754</td>
      <td>71.0</td>
      <td>0.746583</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTTGACTGAGGCGA-1</th>
      <td>mEB_day4</td>
      <td>2446</td>
      <td>6908.0</td>
      <td>1999.0</td>
      <td>28.937466</td>
      <td>65.0</td>
      <td>0.940938</td>
    </tr>
    <tr>
      <th>TTTGACTGCATTGG-1</th>
      <td>mEB_day4</td>
      <td>2906</td>
      <td>9558.0</td>
      <td>3067.0</td>
      <td>32.088303</td>
      <td>91.0</td>
      <td>0.952082</td>
    </tr>
    <tr>
      <th>TTTGACTGCTGGAT-1</th>
      <td>mEB_day4</td>
      <td>1475</td>
      <td>3280.0</td>
      <td>1035.0</td>
      <td>31.554878</td>
      <td>22.0</td>
      <td>0.670732</td>
    </tr>
    <tr>
      <th>TTTGACTGGTGAGG-1</th>
      <td>mEB_day4</td>
      <td>2808</td>
      <td>9123.0</td>
      <td>2923.0</td>
      <td>32.039898</td>
      <td>55.0</td>
      <td>0.602872</td>
    </tr>
    <tr>
      <th>TTTGACTGTACAGC-1</th>
      <td>mEB_day4</td>
      <td>3518</td>
      <td>14918.0</td>
      <td>5091.0</td>
      <td>34.126560</td>
      <td>91.0</td>
      <td>0.610001</td>
    </tr>
  </tbody>
</table>
<p>5405 rows × 7 columns</p>
</div></div>
</div>
<p>Now let’s vizualize this using a multi-panel violin plot</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">axs</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_ribo&#39;</span><span class="p">],</span><span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/usr/local/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.
  c.reorder_categories(natsorted(c.categories), inplace=True)
... storing &#39;sampleName&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/day_06_scRNAseq_analysis_42_1.png" class="no-scaled-link" src="../_images/day_06_scRNAseq_analysis_42_1.png" style="width: 1581px; height: 381px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/day_06_scRNAseq_analysis_43_0.png" class="no-scaled-link" src="../_images/day_06_scRNAseq_analysis_43_0.png" style="width: 302px; height: 304px;" />
</div>
</div>
<p>Finally, let’s filter some of these barcodes out based on total_counts, to exclude possible doublets, and to remove “low quality” cells based on a mt and ribo count:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">20000</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">,</span> <span class="p">:]</span>
<span class="c1"># filter for percent ribo &gt; 0.05</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_ribo&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">50</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">,</span> <span class="s1">&#39;pct_counts_ribo&#39;</span><span class="p">],</span><span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/day_06_scRNAseq_analysis_45_0.png" class="no-scaled-link" src="../_images/day_06_scRNAseq_analysis_45_0.png" style="width: 1581px; height: 381px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/day_06_scRNAseq_analysis_46_0.png" class="no-scaled-link" src="../_images/day_06_scRNAseq_analysis_46_0.png" style="width: 313px; height: 300px;" />
</div>
</div>
<p>This is where we finished class. We will pick up here on Tuesday</p>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="05_scRNAseq_data_generation.html" class="btn btn-neutral float-left" title="scRNAseq data generation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="07_scRNAseq_analysis_part_2.html" class="btn btn-neutral float-right" title="scRNAseq analysis part 2" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Patrick Cahan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-LFQLTP4DZG"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-LFQLTP4DZG', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>